src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs

```diff
@@ -2999,23 +2999,53 @@ private void VisitStatementsWithLocalFunctions(BoundBlock block)
             if (!TrackingRegions && !block.LocalFunctions.IsDefaultOrEmpty)
             {
                 // First visit everything else
+                var queue1 = ArrayBuilder<BoundLocalFunctionStatement>.GetInstance();
                 foreach (var stmt in block.Statements)
                 {
-                    if (stmt.Kind != BoundKind.LocalFunctionStatement)
+                    if (stmt is BoundLocalFunctionStatement localFunc)
+                    {
+                        queue1.Add(localFunc);
+                    }
+                    else
                     {
                         VisitStatement(stmt);
                     }
                 }
 
-                // Now visit the local function bodies
-                foreach (var stmt in block.Statements)
+                // Now visit the local function bodies.
+                // We first visit those we have seen called (hence we know their starting state),
+                // repeating this while seeing more call sites inside other local functions,
+                // and only then we visit the completely unreachable bodies.
+                // This avoids unnecessary passes and incorrect starting states.
```

Comment:
>This avoids unnecessary passes [](http://example.com/codeflow?start=19&length=30)

> This avoids unnecessary passes

Is this true? It looks like ```VisitLocalFunctionUse``` forces an extra pass, and it looks like the pass is going to happen no matter what we do here.

Reply:
I believe it is true. `VisitLocalFunctionUse` schedules an extra pass only if the function's body has been already visited.

Reply:
`VisitLocalFunctionUse` schedules an extra pass only if the body has been visited. So visiting bodies after callsites reduces unnecessary passes unless the function has more callsites afterwards.

Reply:
> VisitLocalFunctionUse schedules an extra pass only if the body has been visited. So visiting bodies after callsites reduces unnecessary passes unless the function has more callsites afterwards.

It is quite possible I am missing something here. The way it looks to me:
1. ```VisitLocalFunctionUse``` schedules an extra pass
2. Visiting a local function body does not un-schedule any extra passes

Therefore, it looks to me that order of visiting doesn't really impact amount of passes.

Reply:
```cs
local2();
void local1() { }
void local2()
{
    // ...some code here changing nullability state...
    local1();
}
```

1. We visit `local2();` which doesn't schedule another pass since `localFunctionState.Visited` is `false`.
2. We skip `local1` body analysis since we haven't seen any callsite of it.
3. We visit  `local2` body, see call of `local1();` - that doesn't schedule another pass either since `localFunctionState.Visited` is `false` here too.
4. We analyze `local1` body.

All of that happens in one pass whereas before this PR it would require two passes:

1. We visit `local2();` same as before.
2. We visit `local1` body, this sets its `localFunctionState.Visited` to `true`.
3. We visit `local2` body, see `local1();` call and schedule another pass (since `local1`'s `localFunctionState.Visited` is `true`).